[PyTorch] Add codegen unboxing ability

larryliu0820 opened the pull request:
RFC: https://github.com/pytorch/rfcs/pull/40

This PR (re)introduces python codegen for unboxing wrappers. Given an entry of `native_functions.yaml` the codegen should be able to generate the corresponding C++ code to convert ivalues from the stack to their proper types. To trigger the codegen, run
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen
```

Merged changes on CI test. In #71782 I added an e2e test for static dispatch + codegen unboxing. The test exports a mobile model of mobilenetv2, load and run it on a new binary for lite interpreter: `test/mobile/custom_build/lite_predictor.cpp`. 

## Lite predictor build specifics

1. Codegen: `gen.py` generates `RegisterCPU.cpp` and `RegisterSchema.cpp`. Now with this PR, once `static_dispatch` mode is enabled, `gen.py` will not generate `TORCH_LIBRARY` API calls in those cpp files, hence avoids interaction with the dispatcher. Once `USE_LIGHTWEIGHT_DISPATCH` is turned on, `cmake/Codegen.cmake` calls `gen_unboxing.py` which generates `UnboxingFunctions.h`, `UnboxingFunctions_[0-4].cpp` and `RegisterCodegenUnboxedKernels_[0-4].cpp`.
2. Build: `USE_LIGHTWEIGHT_DISPATCH` adds generated sources into `all_cpu_cpp` in `aten/src/ATen/CMakeLists.txt`. All other files remain unchanged. In reality all the `Operators_[0-4].cpp` are not necessary but we can rely on linker to strip them off.

## Current CI job test coverage update

Created a new CI job `linux-xenial-py3-clang5-mobile-lightweight-dispatch-build` that enables the following build options:
* `USE_LIGHTWEIGHT_DISPATCH=1`
* `BUILD_LITE_INTERPRETER=1`
* `STATIC_DISPATCH_BACKEND=CPU`

This job triggers `test/mobile/lightweight_dispatch/build.sh` and builds `libtorch`. Then the script runs C++ tests written in `test_lightweight_dispatch.cpp` and `test_codegen_unboxing.cpp`. Recent commits added tests to cover as many C++ argument type as possible: in `build.sh` we installed PyTorch Python API so that we can export test models in `tests_setup.py`. Then we run C++ test binary to run these models on lightweight dispatch enabled runtime.

larryliu0820 commented:
Based on feedback from PR #68126

iseeyuan reviewed (commented):
Would be great to add some tests: input as a schema (with existing unboxed kernel implementation), the output is the generated boxed wrapper. Call the boxed wrapper to make sure the result is the same with the unboxed call.

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -1293,6 +1375,18 @@ def key_func(fn: NativeFunction) -> str:
 
     cpu_fm.write('Functions.cpp', lambda: {})
 
+    if options.enable_codegen_unboxing:
```
It sounds like the idea of the current PR is that:
- we add this flag to optionally generate unboxing codegen, and only use it for mobile
- torchscript (server-side) won't use the unboxing codegen at all (but it would in a later PR? And then the unboxing codegen part would get generated all of the time, so we wouldn't want the flag?)

The main risk that I see is that the unboxing codegen lands without getting tested outside of mobile. So I think a separate PR is fine if it's easier to split up that way, but it would be good to land the changes all at once, including  usage in torchscript.

larryliu0820 committed:
[PyTorch] Add codegen unboxing ability (#69881)

Summary:
This PR (re)introduces python codegen for unboxing wrappers. Given an entry of `native_functions.yaml` the codegen should be able to generate the corresponding C++ code to convert ivalues from the stack to their proper types. To trigger the codegen, run
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/69881

Test Plan:
See code is generated using
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```
Will add unit test and CI in later diffs.

Differential Revision: D33692299

Pulled By: larryliu0820

fbshipit-source-id: 731a50befa34a1fd1533ed9b6f518adbc41f71cd

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,282 @@
+# ivalues from stack to correct arguments to the unboxed kernel, based on corresponding JIT schema. This codegen is
+# an alternative way to generate unboxing wrappers similar to the existing C++ metaprogramming approach but gets the
+# job done statically. These generated unboxing wrappers will be useful under the scenario where we need to register
+# a fixed set of operators known at compile time and thus can save some time in runtime initialization phase.
```
I don't think "needing to register a fixed set of operators known at compile time" is the main reason for this - since you can already do that today with the templated unboxing, through selective build.

My understanding is that the codegen'd unboxing gives the mobile team more flexibility to extend it in various ways - decoupling it from the dispatcher and customizing it for mobile, e.g. by swapping out `IValue` for another (cheaper) boxed representation in the future.

larryliu0820 replied:
Yep I agree. I didn't understand dispatcher too well by the time I wrote it, will update it

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,173 @@
+
+    template_dir = os.path.join(options.source_path, "templates")
+
+    def make_file_manager(install_dir: str) -> FileManager:
```
nit: move this into `tools/codegen/utils.py` so we can re-use it between here and `gen.py`

larryliu0820 committed:
[PyTorch] Refactor make_file_manager so that it can be reused between gen.py and gen_unboxing.py
Also modified docs for `unboxing.py` to better reflect the benefit of having codegen unboxing.

larryliu0820 committed:
[PyTorch] Fix lint

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,169 @@
+
+    @method_with_native_function
+    def __call__(self, f: NativeFunction) -> str:
+        sig_group = CppSignatureGroup.from_native_function(
```
is this used anywhere? It looks like it's just passed directly into `generate_unboxed_kernel_call` further down.

larryliu0820 replied:
Are you asking about `sig_group`? Yes we only get `sig` from it and pass to `generate_unboxed_kernel_call`.

larryliu0820 replied:
So maybe move it into the `else` block?

larryliu0820 replied:
Inlined `generate_unboxed_kernel_call` into `gen_unboxing.py`.

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,169 @@
+
+            # parse arguments into C++ code
+            arguments = unboxing.convert_arguments(
+                args, f.func.arguments.tensor_options
```
Why do you need to deal with / worry about TensorOptions?

Since IValues don't store TensorOptions, and the C++ frontend API has versions of each op that don't take in TensorOptions, can you get away with just converting each argument from an IValue directly without having to:

(a) pay the runtime cost of wrapping everything in a `TensorOptions` object
(b) deal with added complexity in the codegen :)

larryliu0820 replied:
Good point, I merely removed the docs about `TensorOptions` but didn't remove the code. I'll do it.

larryliu0820 committed:
[PyTorch] Remove code to generate TensorOptions
Pass all the values directly to C++ API

larryliu0820 committed:
[PyTorch] Remove unused imports

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,282 @@
+        assert self_arg is not None, "No self argument"
+        arg_list = arg_connector.join(
+            [
+                arguments[a.name].val_name
```
I was trying to figure out why `arguments` needs to be a `Dict[str, ArgumentCppCode]` (and why the `ArgumentCppCode` class is necessary) and I think there are some existing abstractions in the codegen that should help to make this re-use patterns that the existing codegen uses.

This code is trying to call into a C++ API function (e.g. `at::add(at::Tensor& self, at::Tensor& other, at::Scalar& alpha);), and there are a bunch of local variables in scope that were created by unboxing the ivalues.

We have a `translate()` API that basically let's you do just that (lives and is documented here: https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/api/translate.py#L53).

It involves calling a function that has some target schema (in the above example, it's `Tensor& self`, `Tensor& other` and `Scalar& alpha`), while you have some context of local bindings that are in scope (those would be the temporaries that you constructed in the function from the IValues). `translate()` that returns the list of expressions that you need to pass into the function.

You can see an example of how it's used in a few places, e.g. [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/dest/register_dispatch_key.py#L616).

Doing that should also force you to re-use more of the existing data model in codegen, instead of making new classes:

(1) You'll need to create a `List[Binding]` corresponding to the variables that are in scope when you're ready to call the function (example [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/dest/register_dispatch_key.py#L581)).

To do (1), I think you'll want to have a function that operators on an individual `Argument`, and returns a tuple of:
- the code string that's used to unboxed that argument
- The `Binding` object that corresponds to the type of that unboxed argument, with the new (temporary variable) name that it was given.

There's an example piece of codegen that does that for the functionalization pass [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/gen_functionalization_type.py#L43). If you look at where that function is used, you can see that the context it returns is fed directly into a translate call.

(2) the target bindings just correspond to the arguments in the faithful C++ API, since that's what you're calling. The `CppSignature` object represents the function that you're trying to call, so you can use it to get both the target bindings and the actual function call. Something like:
```
TORCH_API void {f.func.name.unambiguous_name()}(Stack & stack) {{
    ...
    at::{faithful_cpp_sig.name()}({', '.join(e.expr for e in translate(your_context, faithful_cpp_sig.arguments()))});
    ....
}}
```
Another advantage of doing it that way is that more of the "boilerplate" string manipulation can all live in the same place, so you don't have to go hunting around for different functions to piece together what the structure of the full codegen output looks like.

larryliu0820 replied:
Thanks for the detailed explanation. Let me give it a try!

larryliu0820 replied:
This is resolved.

bdhirsh commented:
Looks good so far! I just left a pretty large piece of feedback around refactoring parts of the codegen to re-use some existing machinery, so let me know if you want to talk more about it.

larryliu0820 committed:
[PyTorch] Refactor using translate to generate argument list

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Optimize import

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,225 @@
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, ival: str, arg_name: ArgName, *, mutable: bool = False) -> Tuple[
+    Expr, List[str]]:
```
I think it's a little confusing that this function takes in an `ival: str` and a separate `arg_name: ArgName`. I see now that `ival` corresponds to C++ code with the transformation to grab the ivalue to the stack, and `arg_name` is the name of the original variable in the native function schema.

I think there are a couple ways we could make that clearer. One would be to assign the ivalues to temporary variables that we can refer to directly in the codegen, like:
```
TORCH_API void add_Tensor(Stack & stack) {
    auto self_ivalue = std::move(peek(stack, 0, 3));
    at::Tensor self_base = self_ivalue.to<at::Tensor>();
    ...
}
```

Then in the codegen you could have two functions:
- one for unwrapping the ivalue (returning both the code-string to unwrap, and a `Binding` corresponding to the unwrapped value)
- another for unboxing the individual IValue (aka this function). Then this function can just take in the Binding and nothing else (it can figure out `Type` and `mutable` from `Binding.argument`)

------

Another related idea: what do you think of manually writing the functions that unbox each type directly in the template, like `unbox_TensorList`, `unbox_OptionalTensorList`, etc? That seems nice because any function that lives directly in C++ is way easier to reason about than one that's indirectly generated from python codegen. Then this function becomes a lot simpler - Given a `Binding` with some `name`, the code string it spits out is just a single function call (and you don't have to deal with doing any of the recursive calls that you have here for container types, etc).

larryliu0820 replied:
> manually writing the functions that unbox each type directly in the template, like unbox_TensorList, unbox_OptionalTensorList, etc? 

That is a bit overwhelming since we have more than 20 types in native_functions.yaml, also it cannot handle a type that is not known from native_functions.yaml

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,225 @@
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
```
nit: we should be able to define `out_name` independently of the type (and avoid having to do it separately in a bunch of different elif cases)? Maybe just `f'{original_arg_name}_from_ivalue'` or something.

larryliu0820 replied:
So the problem is I have a recursion and the name need to be different in each recursive call.

bdhirsh replied:
https://github.com/pytorch/pytorch/pull/69881#discussion_r793008698 would help with that?

bdhirsh commented on aten/src/ATen/templates/CodegenUnboxingWrappers.cpp:
```diff
@@ -0,0 +1,42 @@
+
+// ${generated_comment}
+
+// NOTE [Sharded File]: This file is generated in a sharded fashion to speed up
```
You can probably kill this comment and replace it with `See Note [Sharded File]`

Can you also add a comment about this file? We probably want to explicitly mention that we're code-generating registrations to the JIT registry here, but they're currently only used by mobile builds.

bdhirsh commented on aten/src/ATen/templates/CodegenFunctions.h:
```diff
@@ -0,0 +1,30 @@
+// ${generated_comment}
```
nit: can we make the name less generic, maybe something like `UnboxingFunctions.h` (similar to how we have a `Functions.h` for the normal unboxed C++ API).

and can you add a comment at the top of the file describing the purpose of the API? (Something about how we code-generate an unboxing API for every native operator).

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+        for a in sig.arguments()
+    )
+    # find dispatch native function namespace
+    if use_tensor_options:
```
i still think we can get away without this: is there anything in the `torch` namespace that you actually need? You should be able to get away with using the `at::` namespace for everything here (since the output file also lives in `aten`)

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+    arg_list = f"\n\t{arg_connector.join(e.expr for e in expr_list)}\n"
+    if Variant.method in f.variants:
+        function_call = f"""
+{ret_str}self_base.{sig.name()}({arg_list});
```
I think we can even remove this function completely, and move it all directly into `ComputeUnboxingFunctions.__call__()`. The more individual `f-strings` you have, and the more scattered they are across different files, the harder it is to see what the final output of the codegen is. Maybe more of this can be merged into a single f-string, that you can look at and get a better idea of what the whole output is.

bdhirsh replied:
(the translate() call and the actual string that calls the C++ API are the only places where I think the `CppSignature` is used, so I think it would also make that clearer).

larryliu0820 commented on aten/src/ATen/templates/CodegenUnboxingWrappers.cpp:
```diff
@@ -0,0 +1,42 @@
+#include <torch/csrc/jit/runtime/operator.h>
```
How about `RegisterCodegenUnboxedKernels.cpp`? This thing can be used in embedded which is not limited to mobile.

larryliu0820 committed:
[PyTorch] Address comments, mostly refactoring

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Fix lint again

larryliu0820 committed:
[PyTorch] Fix unused import

larryliu0820 committed:
[PyTorch] Rename function name and modify a few comments

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+        code = _gen_code_list_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    else:
+        raise Exception(f"Cannot handle type {t}. arg_name: {arg_name}")
+    expr = Expr(expr=out_name, type=nctype)
```
Oh I see why you had to add the extra rule to `translate()` - right here, you're manually creating the `Expr` object.

You could manually write logic here like `Expr(expr=out_name if out_name != "memory_format" else SpecialArgName.possibly_redundant_memory_format, ...)`

...But I don't think you should actually be manually creating the `Expr` object at all (which will end up being fragile, since you have to create it exactly the same way that the C++ api codegen does it)

Instead, you should rely on the C++ api codegen to do it for you (which you can see happens [here](https://github.com/pytorch/pytorch/blob/09e54ffec3d01d2742b7bad5b5cbd5ffc56bcf33/tools/codegen/api/cpp.py#L258)). We already have a way for generating those bindings, through the `CppSignature` object. (you can see the functionalization codegen doing something similar [here](https://github.com/pytorch/pytorch/blob/63429bf4b3f491164f9896701134d81162b27469/tools/codegen/gen_functionalization_type.py#L52))

One way would be to do something like:
```
# pass this CppSignature(...).most_faithful_signature()
def convert_arguments(sig: CppSignature) -> Tuple[List[Binding], List[str]]:
    code_list = []
    binding_list = []
    args = sig.arguments()
    for i, a in enumerate(args):
        # get the code for unwrapping the ivalue, and the name of the variable we unwrapped
        unboxed_arg_name, code = argumenttype_ivalue_convert(a.argument.type, a.name, mutable=a.argument.is_write)
        code_list.append(f"""
c10::IValue {a.name} = std::move(peek(stack, {i}, {len(args)}));
{'\n'.join(block for block in code)}
        # add the binding for our new unwrapped variable to the context
        binding_list.append(a.with_name(unboxed_arg_name)
        ...
```

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+        out_name = arg_name + "_opt_out"
+        code = _gen_code_optional_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    elif isinstance(t, ListType):
+        out_name = arg_name + "_list_out"
```
nit: switch to f-string

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+        out_name = f"{arg_name}_base"
+        code = [f"{ctype} {arg_name}_base = {arg_name}.to<{ctype}>();"]
+    elif isinstance(t, OptionalType):
+        out_name = arg_name + "_opt_out"
```
nit: switch to f-string

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[Expr, List[str]]:
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
```
In general we want to keep everything as a type in our type system for as long as possible, and only convert to string at the very end (for example when I see `_gen_code_optional_type` takes in a `string`, I have to scroll up here to see that it actually came from a `CType` object).

Can you just plumb around the `CType` directly, and call `.cpp_type()` on it directly in all of the f-strings where it gets used?

You can also plumb the underlying `CType` around instead of the `NamedCType`, since the name info is irrelevant in all of these sites. E.g. something like:
```
ctype: CType = cpp.argumenttype_type(t=t, mutable=mutable, binds=arg_name).type
```

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
+        expr, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
```
nit: I think that generally the fewer strings you have floating around the codegen, the easier it is to reason about things. Instead of the separate lists for `code_list` and `pop_ivalue`, can you do something like
```
code_list.append(f"""
c10::IValue {arg_name} = std::move(peek(stack, {i}, {len(args)}));
{'\n'.join(block for block in code)}
""")
```

bdhirsh replied:
The new changes are still pretty non-local. What's nice about the code snippet I put in the above comment are there all of the pieces of code are in the same f-string, so very localized and easy to see why they fit together. ("I create an ivalue named `{arg_name}`, so clearly the next line will use it").

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+    pop_ivalue = []
+    code_list = []
+    for i, arg in enumerate(args):
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
```
nit: I would just move the argument string inside of the loop and use an f-string, that way you know exactly what the parameters to it are at it's creation site.

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
```
nit: TensorOptions not needed anymore (in the comment)

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+from typing import List, Tuple
+
+from tools.codegen.api.cpp import argumenttype_type
```
nit: we have a whole bunch of `argumenttype_type` functions that are all under different api's (`dispatcher.argumenttype_type`, `python.argumenttype_type`, etc).

Can you just `from tools.codegen.api import cpp`, that way you have to use `cpp.argumenttype_type` directly and it's clear which api it is?

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,192 @@
+def _gen_code_list_type(arg_name: str, out_name: str, t: ListType, ctype: str) -> List[str]:
+    in_name = arg_name + "_list_in"
+    elem_name = arg_name + "_elem"
+    code = [f"const c10::List<c10::IValue> {in_name} = {arg_name}.toList();"]
```
Hmm... I think this is probably okay

We generally want to avoid hardcoding C++ types as strings into random places in the codegen, so there's one source of truth for everything. There's already a notion of "The C++ type for list" in `tools/codegen/api/types.py`. But that file doesn't know what `IValue`'s are, and I'm not sure that we should add that yet. No action necessary though in this PR.

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,177 @@
+            args_str = "" if not expr_list else f"""
+        {arg_connector.join(e.expr for e in translate(expr_list, sig.arguments(), method=sig.method))}
+    """
+            ret_type: CType = cpp.returns_type(f.func.returns)
```
nit: most pattern matching on aspects of the schema should use the `FunctionSchema` directly, since you don't actually care about the C++ specific details here. You can just check e.g. if `len(f.func.returns) == 0` (and you don't care about what that implies for the C++ return type).

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,177 @@
+            f, method=(Variant.method in f.variants), fallback_binding=f.manual_cpp_binding
+        )
+
+        sig = sig_group.signature
```
is there a reason this isn't `most_faithful_signature()`? I would've thought you'd end up with the versions of ops that takes `TensorOptions` if you used this. (you can probably sanity check by looking at the generated `RegisterCodegenUnboxedKernels.cpp`)

larryliu0820 replied:
Don't think it matters since we are only using the `sig.name()` but will align with the other use case.

bdhirsh reviewed (changes requested):
Left another round of feedback, I think codegen is coming along.

Another thing that I think should happen in this PR - we want mobile interpreter tests to ensure that they cover every type expressed in the codegen'd unboxing logic. Have you checked if we're covering all the cases there?

larryliu0820 committed:
[PyTorch] Generate binding list instead of expr list for arguments

larryliu0820 commented:
>  we want mobile interpreter tests to ensure that they cover every type expressed in the codegen'd unboxing logic

So I'm assuming you are referring to the end-to-end tests we talked about earlier for mobile interpreter. The thing is that only comes after we hookup the generated source files into the build system (cmake/bazel). This PR doesn't have it yet since it's another big change #71782. Do you think it's better if we merge these two PRs into one and add tests as well?

larryliu0820 committed:
[PyTorch] Address part of the comments

iseeyuan commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -30,6 +30,9 @@ TORCH_LIBRARY_IMPL(aten, Conjugate, m) {
   m.impl("conj_physical_", torch::CppFunction::makeFallthrough());
   m.impl("resolve_conj", torch::CppFunction::makeFallthrough());
   m.impl("resolve_neg", torch::CppFunction::makeFallthrough());
+  m.impl("repeat_interleave.Tensor", torch::CppFunction::makeFallthrough());
```
Any reason we need to move these registrations here?

iseeyuan commented on aten/src/ATen/FuncTorchTLS.h:
```diff
@@ -25,6 +25,11 @@ namespace at { namespace functorch {
+  // functorch doesn't always work with autograd.Function.
+  // This is a hook to get into functorch -- functorch will determine
+  // if it should raise an error message
+  virtual int64_t checkSupportsAutogradFunction() const = 0;
```
Why the code-gen is related to FuncTorch?

larryliu0820 replied:
Yeah I rebased and messed up with the view. Trying to fix it

iseeyuan commented on aten/src/ATen/core/dynamic_type.cpp:
```diff
@@ -271,6 +277,16 @@ TypePtr DynamicType::fallback() const {
       return VarType::create(*name_);
     case Tag::AnyClass:
       return AnyClassType::get();
+    case Tag::QScheme:
```
Ah, looks like it's rebased changes. Don't know why it shows up here. Is there any way to filter them out for easier review?

larryliu0820 committed:
[PyTorch] Address the rest of comments

larryliu0820 committed:
[PyTorch] Add return type in function signature

larryliu0820 committed:
[PyTorch] Remove unused imports

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,176 @@
+
+            # parse arguments into C++ code
+            binding_list, code_list = convert_arguments(args,
+                                                        has_tensor_options=f.func.arguments.tensor_options is not None)
```
nit: I still think we can get away with zero references to `tensor_options` :) Why do we need this flag?

Separately, can `convert_arguments` just take in `f.func` directly? I think of `convert_arguments` as "given this function, spit out code and bindings for the newly unboxed arguments of that function". As a caller, I don't have to care about exactly which arguments in the schema to pass in.

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,199 @@
+from typing import List, Tuple
+
+from tools.codegen.api.cpp import argumenttype_type, argument
```
See my previous comment here: https://github.com/pytorch/pytorch/pull/69881/#discussion_r794867257.

we have a bunch of `def argument()` functions floating around in the codegen under different API's, so you want to be forced to call `cpp.argument()` in this file, instead of just calling `argument()`. That way it's immediately clear which API is in use.

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,199 @@
+        code_list.extend(code)
+        # we need the 'self' argument so method needs to be False, also it doesn't matter if signature faithful or not
+        arg_binding = argument(arg, cpp_no_default_args=set(), method=False, faithful=False,
+                               has_tensor_options=has_tensor_options)
```
Okay, so we don't really care about half of the inputs to this function. That's a sign to me that we should be using higher-level api's to do what we're trying to do.

you could directly use `cpp.arguments()` instead, but that also deals directly with `has_tensor_options` and `cpp_no_default_args` stuff.

You can see that the `CppSignature` class abstracts some of that away though, here: https://github.com/pytorch/pytorch/blob/72c972e1e1b4ad838de604e35269e200a70db5f2/tools/codegen/api/types.py#L422. You can pass in the NativeFunction and nothing else, and you'll get a `CppSignature` with all of the correct "default" properties for that NativeFunction (including the fact that say, `cpp_no_default_args` was generated from `f.cpp_no_default_args`.

So when I see that, it tells me that maybe this `convert_arguments()` function should just directly take in the entire `NativeFunction`. Then you can do something like:
```
def convert_arguments(f: NativeFunction) -> Tuple[List[Binding], List[str]]:
    cpp_args = CppSignature.from_native_function(f).most_faithful_signature().arguments()
    binding_list = []
    code_list = []
    for i, arg in enumerate(cpp_args):
        unboxed_name, code = argumenttype_ivalue_convert(arg.argument.type, arg.argument.name, mutable=arg.argument.is_write)
        # generate code for unboxing each ivalue
        code_list.append(f"""\n
c10::IValue {cpp_args[i].name} = std::move(peek(stack, {i}, {len(cpp_args)}));
{code}"""
      
        binding_list.append(arg.with_name(unboxed_name))
```

bdhirsh replied:
Or alternatively, you could probably directly pass the `CppSignature` instead of the `NativeFunction` since it carries all of the same information, and you need to re-use the `CppSignature` in `gen_unboxing` later on

larryliu0820 replied:
I'll need to pass a `NativeFunction` since I need to set `method=False` for the `CppSignature`. This makes sure the self argument is present.

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,199 @@
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[str, CType, List[str]]:
```
Do we actually need to return the `CType` in this function? It looks like it's only used in one place and ignored everywhere else.

Why not just compute it directly in the one place where it's needed? (in `_gen_code_list_type`)

bdhirsh replied:
On a similar note, can you add in the comment above this file what everything that it returns corresponds to?

It would be kind of nice if we could just take in an `Argument` instead of the separate componeents of it (type / name / mutability), but I guess you need that in order to make it recursively work on nested types.

"Takes in the type, name and mutability corresponding to an argument, and generates a tuple of:
(1) the C++ code necessary to unbox the argument
(2) A `Binding` corresponding to the newly created unboxed variable
"

bdhirsh replied:
The `CType` is still being returned here? Why do we need it? (see above comment - can't we just generate outside of the function in one place where its needed)

larryliu0820 replied:
We need `CType` for list element. If we generate outside of the function it will only be a type of List but not the type of the element.

bdhirsh replied:
Ok. Mechanically it looks like you can't because you don't have e.g the `is_mutable` info by the time you're in the list function.

I think this is fine good enough shape for now. I could imagine another way of doing this by passing the `Argument` object around everywhere, but you might run into some issues doing that too. No more action needed :)

bdhirsh commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,176 @@
+    @method_with_native_function
+    def __call__(self, f: NativeFunction) -> str:
+        sig_group = CppSignatureGroup.from_native_function(
+            f, method=(Variant.method in f.variants), fallback_binding=f.manual_cpp_binding
```
nit 1: the signature is only used in the `else` clause, so may as well wait until you're inside the `else` clause to generate it

nit 2: you don't care about `manual_cpp_binding` in the unboxing code and it has a default value anyway, so may as well not pass it in

larryliu0820 commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,176 @@
+            arg_connector = ",\n\t\t"
+            # function call and push back to stack
+            prefix = "self_base." if sig.method else "at::"
+            args_str = "" if not binding_list else f"""
```
This is to make the generated code a bit more elegant. Instead of having something like
```
auto result_ = self_base.arccos_(
        
);
```
We can have:
```
auto result_ = self_base.arccos_();
```

bdhirsh replied:
Oh, I'd argue that instead of `arg_connector` having newlines between each arg, you should just put them all on the same line like most other codegen function call-sites. Then you can also remove special-casing logic like that from the codegen

larryliu0820 committed:
[PyTorch] Address comments

larryliu0820 commented:
@bdhirsh I'll rebase #71782 on top of my newest commit of this PR and enable end-to-end test there.

larryliu0820 committed:
[PyTorch] Fix lint in unboxing.py

larryliu0820 committed:
[PyTorch] Remove unused import

bdhirsh commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def _gen_code_list_type(arg_name: str, out_name: str, t: ListType, ctype: CType) -> List[str]:
+    ctype_str = ctype.cpp_type(strip_ref=True)
```
nit: in the spirit of "avoid making strings until the last possible second", I'd just put `ctype.cpp_type(strip_ref=True)` directly inside of the f-string

larryliu0820 committed:
[PyTorch] Address comments inline cpp type str, cleanup comments, put arguments into one line

larryliu0820 committed:
[PyTorch] Add space between arguments

larryliu0820 committed:
[PyTorch] Enable lightweight dispatch as an option in cmake build (#71782)

* [PyTorch] Add codegen unboxing ability (#69881)

Summary:
This PR (re)introduces python codegen for unboxing wrappers. Given an entry of `native_functions.yaml` the codegen should be able to generate the corresponding C++ code to convert ivalues from the stack to their proper types. To trigger the codegen, run
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/69881

Test Plan:
See code is generated using
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```
Will add unit test and CI in later diffs.

Differential Revision: D33692299

Pulled By: larryliu0820

fbshipit-source-id: 731a50befa34a1fd1533ed9b6f518adbc41f71cd

* [PyTorch] Refactor make_file_manager so that it can be reused between gen.py and gen_unboxing.py
Also modified docs for `unboxing.py` to better reflect the benefit of having codegen unboxing.

* [PyTorch] Fix lint

* [PyTorch] Remove code to generate TensorOptions
Pass all the values directly to C++ API

* [PyTorch] Remove unused imports

* [PyTorch] Refactor using translate to generate argument list

* [PyTorch] Fix lint

* [PyTorch] Optimize import

* [PyTorch] Address comments, mostly refactoring

* [PyTorch] Fix lint

* [PyTorch] Fix lint again

* [PyTorch] Fix unused import

* [PyTorch] Rename function name and modify a few comments

* [PyTorch] Generate binding list instead of expr list for arguments

* [PyTorch] Address part of the comments

* [PyTorch] Address the rest of comments

* [PyTorch] Add return type in function signature

* [PyTorch] Remove unused imports

* [PyTorch] Address comments

* [PyTorch] Fix lint in unboxing.py

* [PyTorch] Remove unused import

* [PyTorch] Address comments inline cpp type str, cleanup comments, put arguments into one line

* [PyTorch] Add space between arguments

* [PyTorch] Enable lightweight dispatch as an option in cmake build

* [PyTorch] Fix lint

* [PyTorch] Enable lightweight dispatch as an option in cmake build

* [PyTorch] Add support for bazel build

* [PyTorch] Enable lightweight dispatch as an option in cmake build

* [PyTorch] Fix BUILD.bazel

* [PyTorch] Fix bazel

* [PyTorch] Retry fixing BUILD.bazel

* [PyTorch] Remove merge conflict

* [PyTorch] Add trailing newline

* [PyTorch] Incorporate #71504 changes

Avoid registering into dispatcher but only register into JIT op registry. This commit enables the CI linux-xenial-py3-clang5-mobile-custom-build-static test with lightweight dispatch enabled.

* [PyTorch] Fix codegen.cmake

* [PyTorch] Fix codegen.cmake by setting sources conditionally

* [PyTorch] Modify static dispatch CI test to support codegen unboxing

* [PyTorch] Fix cmake lint

* [PyTorch] Enable both mobile and server build CI tests

* [PyTorch] Too many jobs failing, reverting Bazel file changes to debug

* [PyTorch] Fix lint

Co-authored-by: Larry Liu <8188269+larryliu0820@users.noreply.github.com>

iseeyuan commented:
> Merged changes on CI test. In #71782 I added an e2e test for static dispatch + codegen unboxing. The test exports a mobile model of mobilenetv2, load and run it on a new binary for lite interpreter: `test/mobile/custom_build/lite_predictor.cpp`.
> ...

It's very informative description. Would be great to move it to the summary.

iseeyuan commented on cmake/Codegen.cmake:
```diff
@@ -268,13 +304,16 @@ if(INTERN_BUILD_ATEN_OPS)
 
+  if(NOT USE_LIGHTWEIGHT_DISPATCH)
+    set(generated_unboxing_sources "")
+  endif()
```
Could we add else, to empty dispatching sources with TORCH_LIBRARY? We could generate both, but would the compiler time be longer?

larryliu0820 replied:
Cool I think I can merge this to line 221

iseeyuan commented on tools/codegen/api/unboxing.py:
````diff
@@ -0,0 +1,201 @@
+#       Generates C++ code to convert an ivalue (from stack) to its underlying C++ type.
+#    - int[] size
+#        ```cpp
+#           const c10::List<c10::IValue> size_list_in = (std::move(peek(stack, 0, 7))).toList();
````
Where's 0 and 7 from? How do we know the which locations from the stack are the list elements?

larryliu0820 replied:
`gen.py` or rather the whole codegen framework provided all these information in Python APIs. Fundamentally the source of truth is `native_functions.yaml`.

iseeyuan commented on tools/codegen/api/unboxing.py:
````diff
@@ -0,0 +1,201 @@
+#                                 ~~~~~~~~~~~~~ <-- The converted argument from ivalues in the stack.
+#                                                   Will be passed to unboxed kernel.
+#       ```
+#    - Dimname[]? names
````
I assume it's for completeness, but there's no use case of Dimname[] in Edge afaik.

larryliu0820 replied:
Right but maybe in the future? The outcome of this PR is that we are able to codegen all the unboxing kernels for operators in `native_functions.yaml`.

ezyang replied:
nah, Dimname is pretty dead, we're more likely to remove it than push it into edge

iseeyuan commented:
It's for CMake. Are you planning to change the internal BUCK files for the codegen?

larryliu0820 commented:
So I will definitely enable it in internal BUCK (in diffs). That work is entangled with tracing based selective build support, needs more time.

I'm also going to add more tests similar to [test/cpp/jit/test_lite_interpreter.cpp](https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_lite_interpreter.cpp).

larryliu0820 committed:
[PyTorch] Refactor Codegen.cmake to have better logic on USE_LIGHTWEIGHT_DISPATCH option

larryliu0820 committed:
[PyTorch] Remove unnecessary colon

iseeyuan reviewed (approved):
LGTM. @bdhirsh may take a final look before it's landed.

iseeyuan commented:
> I'm also going to add more tests similar to [test/cpp/jit/test_lite_interpreter.cpp](https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_lite_interpreter.cpp).

It would be great to add unit tests. Just wondering, if the CMake switch is turned on, would the existing unit tests be using the codegen in this PR? Then we have good coverage, and only need to add a test job with the switch turned on.

bdhirsh commented:
Codegen looks good to me!

I think before landing it, we want to ensure there are mobile tests that cover each of the main code paths in the codegen unboxing. For example, someone might try to run a mode in mobile later, using an op that takes in a List[Optional[Tensor]] and run into hard-to-track-down bugs that end up getting diagnosed to a codegen bug (or non-mobile, if non-mobile workflows ever use this codegen). We can pre-empt that by going through the set of mobile interpreter C++ tests and ensuring that there's at least one op called that has a schema with arguments matching the different arg types in this PR.

ezyang commented:
Hi, I haven't been following the ongoing conversation as I was on leave. Is there an associated design document for this PR? The PR description is a bit sparse.

larryliu0820 commented:
@ezyang Hi Edward, here's the design doc: https://docs.google.com/document/d/15VXyruwjr1VFe_FIzNzo0pg3kZaqAWM6DWZTqnnYVnk/edit#heading=h.18gz2zad8lz8.  Let me know if you have any questions!

ezyang commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,201 @@
+# This file generates the code for unboxing wrappers, i.e., the glue logic to unbox a boxed operator and convert the
+# ivalues from stack to correct arguments to the unboxed kernel, based on corresponding JIT schema. This codegen is
+# an alternative way to generate unboxing wrappers similar to the existing C++ metaprogramming approach but gets the
+# job done statically. These generated unboxing wrappers will be useful under the scenario where we need to register
```
* How are we validating that the metaprogrammed templates are equivalent to the code generated unboxing wrappers here?
* If I change some logic here, how does it have to be reflected in the C++ metaprogram? (And vice versa?)

larryliu0820 replied:
> How are we validating that the metaprogrammed templates are equivalent to the code generated unboxing wrappers here?

I'm adding tests to cover the types for mobile use cases.

> If I change some logic here, how does it have to be reflected in the C++ metaprogram? (And vice versa?)

These two are not coupled right now. We didn't think of the need to couple them together, since this is used by *mobile only* and completely separated from server runtime.

ezyang replied:
> These two are not coupled right now. We didn't think of the need to couple them together, since this is used by mobile only and completely separated from server runtime.

This is an interesting claim. While I certainly agree that the modification here affords some amount of decoupling (e.g., if server adds a problematic feature that mobile doesn't need, codegen unboxing can choose not to support that functionality for now), it seems that to keep our system coherent, we want the C++ metaprogram and the codegen to "do the same thing" and that means keeping both sides of the system synchronized if you make a change. This isn't hypothetical; for example https://github.com/pytorch/pytorch/pull/67964 is the kind of change we're interested in making to the calling convention.

larryliu0820 replied:
I see. I am planning to add the same set of tests to metaprogramming and codegen unboxing, hopefully these tests can serve the purpose of making sure they are synchronized.

larryliu0820 committed:
[PyTorch] Add unit tests for kernel arguments unboxing (WIP)

ezyang commented:
Thanks a lot for submitting the rfc!

larryliu0820 committed:
[PyTorch] Enable test_jit for mobile custom build static CI job (WIP)

larryliu0820 committed:
[PyTorch] Add updated yaml file (WIP)

larryliu0820 committed:
[PyTorch] Add a new workflow with setup.py build for lightweight dispatch (WIP)

larryliu0820 committed:
[PyTorch] Cleanup

larryliu0820 committed:
[PyTorch] Generate workflow yaml

larryliu0820 committed:
[PyTorch] Fix typo and cmake messages

larryliu0820 committed:
[PyTorch] Fix script again

larryliu0820 committed:
[PyTorch] Fix custom_build/build.sh

larryliu0820 committed:
[PyTorch] Fix custom_build/build.sh again

larryliu0820 committed:
[PyTorch] Try something new

larryliu0820 committed:
[PyTorch] Do not use Clang to avoid compile error
I'm seeing build failure at
```
third_party/fbgemm/include/fbgemm/FbgemmFPCommon.h:201:48: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
              std::array<float, 14 * 32> c_tmp{0.f};
```
After some research I believe it's only happing with clang.

larryliu0820 committed:
[PyTorch] Only install mobile test

larryliu0820 committed:
[PyTorch] Retry setup.py build, disabling some options

larryliu0820 committed:
[PyTorch] Avoid building caffe2

ezyang commented on tools/setup_helpers/gen_unboxing.py:
```diff
@@ -0,0 +1,11 @@
+
+import tools.jit.gen_unboxing
+
+tools.jit.gen_unboxing.main()
```
@wconstab has been adding new codegen recently, and he asked me about whether or not he should make a new entrypoint or not. I said that, absent other considerations, it's easier to reuse the existing entrypoint. So I'm wondering if there is a good reason to make a new endpoint for unboxing codegen.

ezyang replied:
One justification I can think of is by having a separate endpoint, you can control whether or not you run the codegen or not. But for the sake of debugging, it's far easier if we unconditionally generate all the files in all builds (and then make choices about whether or not to build them or not). Also, if you REALLY wanted, you could still have a flag to toggle whether or not the codegen happened or not.

larryliu0820 replied:
@bdhirsh suggested that this codegen to be completely isolated from `gen.py`. I'm fine with either way.

bdhirsh replied:
Oh, that was mostly my suggestion, mostly for two reasons:

(1) I figured that if mobile is the only consumer of the codegen'd unboxing, then it would be easier to reason about it we keep it in a separate script instead of cluttering up `gen.py` with another flag.

(2) The mobile codegen generates unboxing kernels under the `aten` namespace, but it also generates registrations to the `jit` registry, which live under the `torch` namespace. At least right now, I think that `gen.py` only generates stuff that lives in `aten` (and for example, the autograd codegen has a separate entrypoint. Also I guess that won't be true anymore if/once LTC goes directly in `gen.py`).

wconstab replied:
Thanks for looping me in @ezyang.  My approach (still WIP and only tested via buck so far) is D34333432 if you want to compare.  I think it ends up being pretty convenient to tie into gen.py but I may run into issues as I try to finish it up.

ezyang commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,182 @@
+        depfile_name = depfile_path.name
+        depfile_stem = depfile_path.stem
+
+        path = depfile_path.parent / depfile_name
```
@peterbell10 Do you think there's some refactoring we could have done to reduce the amount of boilerplate code here? This seems like something fm should be able to deal with...

peterbell10 replied:
I don't think so because this code is over-complicated. It's equivalent to:
```python
    if options.output_dependencies:
        depfile_path = pathlib.Path(options.output_dependencies).resolve()
        fm.write_outputs(depfile_path.stem, str(depfile_path))

```
`.resolve()` is optional. `.stem` is the file name minus file extension, and is used as the cmake variable's name.

The over-complication comes from copy-pasting from `gen.py` which is more complicated because it's generating multiple different filenames from one path. But that isn't happing here.

ezyang commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,182 @@
+        sig = sig_group.most_faithful_signature()
+
+        # escape double quote in schema, get rid of extra double quotes
+        schema = json.dumps(sig.func.__str__())[1:-1]
```
Nah, don't do this. Even if it works there is no good reason to use the JSON standard to serialize C++ strings. tools.codegen.gen has an appropriate function to use

```
def cpp_string(s: str) -> str:
    """Convert a python string into a c++ string literal """
    s = s.replace('\\', '\\\\')
    s = s.replace('"', '\\"')
    s = s.replace('\a', '\\a')
    s = s.replace('\b', '\\b')
    s = s.replace('\f', '\\f')
    s = s.replace('\n', '\\n')     
    s = s.replace('\v', '\\v')     
    s = s.replace('\t', '\\t')     
    return f'"{s}"'
```

ezyang commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,182 @@
+    TORCH_SELECTIVE_SCHEMA("aten::{schema}"),
+    [](Stack & stack) {{
+        RECORD_FUNCTION("{sig.name()}", std::vector<c10::IValue>());
+        at::unboxing::{f.func.name.unambiguous_name()}(stack);
```
Instead of directly accessing unambiguous_name(), indirect it through an API module, e.g., `unboxing.name(f)`. Net effect is the same but it is clear that you are getting the unboxing name (and if you need to change the naming convention there is one place you can do it.)

ezyang commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,182 @@
+OperatorGenerator(
+    TORCH_SELECTIVE_SCHEMA("aten::{schema}"),
+    [](Stack & stack) {{
+        RECORD_FUNCTION("{sig.name()}", std::vector<c10::IValue>());
```
Hmm, how come there needs to be a RECORD_FUNCTION here? I don't see it in the "traditional" codepath

larryliu0820 replied:
For mobile side profiling purpose. I can add them in a different PR if you think it's cleaner that way

ezyang replied:
No I don't need it to be in a separate PR, I just need to know what in this path is parity with server side, and what is different. For example, having an xref saying "this needs to get synchronized with that" would go a long way for making sure divergence doesn't happen.

ezyang commented on tools/jit/gen_unboxing.py:
```diff
@@ -0,0 +1,182 @@
+
+# Generates UnboxingFunctions.h & UnboxingFunctions.cpp.
+@dataclass(frozen=True)
+class ComputeUnboxingFunctions:
```
Is there any point in generating the unboxing:: function? It seems to only be called from one site (the OperatorGenerator invocations) and it seems like life would be easier without having to generate these files.

bdhirsh replied:
That was my suggestion too - one nice thing that lets us do is keep all of the unboxing code under the `at::` (and confirms that we don't accidentally use anything in `torch::` to implement it), which would make it a lot easier to re-use the unboxing codegen in core later if we want to.

There's a separate, mobile-only generated file that generates registrations to the JIT registry. Which we only want to build in mobile, since otherwise we would redundantly register stuff to JIT from both that codegen'd file, and from the dispatcher callback registration logic.

We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered - we actually brought that up in one of the Friday syncs, since that would also give this codegen server-side torchscript coverage. But the conclusion was that the main downside of doing so is that we have no way to easily enforce that ops registered to the dispatcher under the `aten` namespace *have* to go through codegen. That conversation might be worth revisiting though.

Separate from that discussion- the extra files are a hassle, so I totally get it if you think it's worth keeping both the unboxing logic and JIT registrations in one file.

ezyang replied:
The layering justification is legitimate; I think my trouble is that I am having trouble thinking of situations where `at::unboxing::` is a useful API, outside of being an API for precisely how it is being used here. A symbol like `at::unboxing::add` is only useful when you know exactly what operator you want (for the JIT interpreter, you need some sort of registry as seen here), and if you know exactly what operator you want, you probably want the unboxed API anyway.

> We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered

I'm not sure I follow this.

bdhirsh replied:
> We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered

This is the thing that we talked a bit about yesterday over VC. We want to avoid registering an op to JIT twice, which seems pretty tricky (and the way that the PR deals with it right now is that it only runs the unboxing codegen when static dispatch is enabled, so we don't generate any dispatcher registrations).

One potential way to do that would be to establish some namespace convention, where e.g. all ops in the `aten` namespace are expected to be registered through codegen while everything else i expected to be registered through the callback. It seems kind of hard to check for that invariant though - there's nothing stopping someone from manually registering an op to the dispatcher under the "aten" namespace outside of the codegen, which feels pretty fragile (I dunno, maybe we could route the codegen'd dispatcher registrations through a different, private API? I'm not sure how much effort it's worth putting into this though).

ezyang commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,201 @@
+#           for (c10::IValue size_elem: size_list_in) {
+#               int64_t size_base = size_elem.to<int64_t>();
+#               size_vec.push_back(size_base);
+#           }
```
@ysiraichi we decided it wasn't doing the https://github.com/pytorch/pytorch/pull/67964 treatment for int lists, right?

ysiraichi replied:
I don't think we discussed about int lists. But, we can add support for them with #69606 changes.

ezyang replied:
I'm a little torn about this because int lists aren't refcounted so copying into a contiguous array isn't too bad; plus people are more likely to repeatedly index into int lists. On the other hand you may be able to avoid a dynamic allocation if the int list size is statically known.

larryliu0820 committed:
[PyTorch] Address comments
Use `gen.cpp_string()` to escape properly. Add a new API in `unboxing` to get unambiguous operator name

larryliu0820 committed:
[PyTorch] Fix caffe2/proto not found issue

larryliu0820 committed:
[PyTorch] Use separate CMakeLists.txt and build script

larryliu0820 committed:
[PyTorch] Fix typo

larryliu0820 committed:
[PyTorch] Change permission to build.sh

ezyang commented on aten/src/ATen/templates/UnboxingFunctions.h:
```diff
@@ -0,0 +1,35 @@
+// function will pop out IValues from the stack then convert them into the correct C++ types based on given schema. This
+// unboxing logic is an alternative to template-based metaprogramming unboxing.
+
+#ifndef PYTORCH_UNBOXINGFUNCTIONS_H
```
Use #pragma once instead?

larryliu0820 committed:
[PyTorch] Test

larryliu0820 committed:
[PyTorch] Finally fix gtest issue

larryliu0820 committed:
[PyTorch] Change permission

larryliu0820 committed:
[PyTorch] Enable test

larryliu0820 committed:
[PyTorch] Add pragma once

larryliu0820 committed:
[PyTorch] Add back a tag for CI job

larryliu0820 committed:
[PyTorch] Regenerate CI job

larryliu0820 committed:
Merge branch 'master' into static_dispatch

larryliu0820 committed:
[PyTorch] Update workflow after merging with master

larryliu0820 committed:
[PyTorch] Run test in build phase since test phase requires dist/whl

larryliu0820 committed:
[PyTorch] Fix tests_setup.py

larryliu0820 committed:
[PyTorch] Add 2 tests

larryliu0820 committed:
[PyTorch] Clean debug logging

larryliu0820 committed:
[PyTorch] Add more tests and fixed a bug on converting optional list

larryliu0820 committed:
[PyTorch] Cleanup

larryliu0820 committed:
Merge branch 'master' into static_dispatch

larryliu0820 committed:
[PyTorch] Include test_lite_interpreter_runtime

ezyang commented on test/mobile/custom_build/CMakeLists.txt:
```diff
@@ -13,12 +13,16 @@ target_include_directories(Predictor PUBLIC ${TORCH_INCLUDE_DIRS})
+if(${APPLE})
+  set(LINKER_OPTIONS "")
+else()
+  set(LINKER_OPTIONS "-Wl,-s -Wl,--gc-sections -Wl,--whole-archive -Wl,--no-whole-archive")
```
This is wrong. `-Wl,--whole-archive` linker flag positioning matters: you have to enclose the objects you want to whole archive with the whole-archive/no-whole-archive; moving them here turns it into a noop

larryliu0820 committed:
[PyTorch] Revert custom_build changes

ezyang commented on test/mobile/lightweight_dispatch/build.sh:
```diff
@@ -0,0 +1,54 @@
+# This script should be called from .jenkins/pytorch/build.sh. Assuming we are at pytorch source root directory.
```
missing a shebang? Also consider `set -ex` (e for quit on error, x for printing commands as they are run).

ezyang commented on test/mobile/lightweight_dispatch/test_codegen_unboxing.cpp:
```diff
@@ -0,0 +1,195 @@
+#include <torch/csrc/jit/frontend/resolver.h>
+#include <torch/csrc/jit/mobile/import.h>
+#include <torch/csrc/jit/mobile/module.h>
+// Cover codegen'd unboxing logic for these types:
```
how to keep this list up to date?

ezyang commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,208 @@
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
```
Just a general remark: the code snippets here seem super wordy. It feels like you could profitably introduce some conversion functions and then make the generated code quite a bit smaller.

larryliu0820 committed:
[PyTorch] Address comments

ezyang commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,208 @@
+
+
+# Convert all the arguments in a NativeFunction to C++ code
+def convert_arguments(f: NativeFunction) -> Tuple[List[Binding], List[str]]:
```
Letting bindings come with prologue `code_list` is something that I have thought could be hypothetically useful, but it's not clear to me that it is useful here. In particular, are you actually using the IValue bindings more than once in the code below? If you are not, you can just do the conversion all in one go (maybe with some helper functions) inside the Binding and then dispense with the code list (which is a little brittle anyway because you have to worry about accidentally using a name multiple times.)

ezyang commented on tools/codegen/api/unboxing.py:
```diff
@@ -0,0 +1,208 @@
+
+
+def _gen_code_base_type(arg_name: str, out_name: str, ctype: CType) -> Tuple[List[str], List[str]]:
+    return [f"{ctype.cpp_type(strip_ref=True)} {out_name} = {arg_name}.to<{ctype.cpp_type(strip_ref=True)}>();"], []
```
This logic (and below) feels like it could profitably live in translate. The idea is we would teach translate how to convert an IValue into a C++ type, and then you could just use translate to get out the Bindings you need for the unboxing.

ezyang commented:
This gets a LGTM from me. There are some changes that can help make this code more modular and usable in other situations, but I will be happy if there is just a comment saying what improvements are possible, they do not have to be done as part of this PR.

larryliu0820 committed:
[PyTorch] Fix broken tests

larryliu0820 committed:
[PyTorch] Add back link options

larryliu0820 committed:
[PyTorch] Try to fix linker options

larryliu0820 committed:
[PyTorch] Fix linker options

larryliu0820 committed:
Merge branch 'master' into static_dispatch

larryliu0820 committed:
[PyTorch] Try GCC instead of Clang

bdhirsh reviewed (approved):
thanks for all of the back and forth!
